### PR TITLE
probe secret

### PR DESCRIPTION
**Pull request by katherinelc321**
This PR probes the osd-admin secret in accounts and heal them if they're invalid https://issues.redhat.com/browse/OSD-9298

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -81,3 +82,16 @@ func GetIAMArn(awsAccountID, awsResourceType, awsResourceID string) (arn string)
 	arn = strings.Join([]string{"arn:", awsAPI, ":iam::", awsAccountID, ":", awsResourceType, "/", awsResourceID}, "")
 	return
 }
+
+func SetHiveSHardName(configMap *corev1.ConfigMap) error {
```
Comment (dkeohane) [inline config/config.go:86]:
```suggestion
func SetHiveShardName(configMap *corev1.ConfigMap) error {
```

File: cmd/manager/main.go [hunk 2]
```diff
@@ -118,6 +118,21 @@ func main() {
 		os.Exit(1)
 	}
 
+	// Define a kubeClient for any processes that need to run during operator startup or independent routines to use
```
Comment (dkeohane) [inline cmd/manager/main.go:121]: From: https://github.com/openshift/aws-account-operator/pull/668#discussion_r810034883
```suggestion
	// Define a kubeClient for any processes that need to run during operator startup or independent routines to use
	// We should avoid using this kubeClient except for when necessary and utilize the operator-sdk provided client as much as possible.
	// The operator-sdk kube client provides a level of caching that we don't get with building our own this way.
```

File: config/config.go [hunk 3]
```diff
@@ -81,3 +82,16 @@ func GetIAMArn(awsAccountID, awsResourceType, awsResourceID string) (arn string)
 	arn = strings.Join([]string{"arn:", awsAPI, ":iam::", awsAccountID, ":", awsResourceType, "/", awsResourceID}, "")
 	return
 }
+
+func SetHiveShardName(configMap *corev1.ConfigMap) error {
```
Comment (iamkirkbater) [inline config/config.go:86]: I don't see where we're calling this.  I think we need to explicitly call this in manager.go first.

**Review by iamkirkbater — CHANGES_REQUESTED** [review]

File: pkg/controller/account/account_controller.go [hunk 4]
```diff
@@ -344,6 +344,11 @@ func (r *ReconcileAccount) Reconcile(request reconcile.Request) (reconcile.Resul
 	}
 	regionAMIs := processConfigMapRegions(stringRegions)
 
+	awsAssumedRoleClient, creds, err := r.handleCreateAdminAccessRole(reqLogger, currentAcctInstance, awsSetupClient)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
```
Comment (iamkirkbater) [inline pkg/controller/account/account_controller.go:351]: Is there a reason this got moved up?  This seems to be breaking STS cluster initialization.

Comment (iamkirkbater) [reply pkg/controller/account/account_controller.go:351]: Ahh, figured it out.  Patch included in review comment.

**Review by iamkirkbater — CHANGES_REQUESTED** [review]
This breaks during STS testing. `make test-sts` was able to catch this.  `make test-integration` runs through all of our integration tests.

Here's a patch of the changes I made to get this working:
```
diff --git a/pkg/controller/account/account_controller.go b/pkg/controller/account/account_controller.go
index 3b1220f4..1975f77b 100644
--- a/pkg/controller/account/account_controller.go
+++ b/pkg/controller/account/account_controller.go
@@ -344,11 +344,6 @@ func (r *ReconcileAccount) Reconcile(request reconcile.Request) (reconcile.Resul
 	}
 	regionAMIs := processConfigMapRegions(stringRegions)
 
-	awsAssumedRoleClient, creds, err := r.handleCreateAdminAccessRole(reqLogger, currentAcctInstance, awsSetupClient)
-	if err != nil {
-		return reconcile.Result{}, err
-	}
-
 	// Account init for both BYOC and Non-BYOC
 	if currentAcctInstance.ReadyForInitialization() {
 		reqLogger.Info("initializing account", "awsAccountID", currentAcctInstance.Spec.AwsAccountID)
@@ -397,6 +392,11 @@ func (r *ReconcileAccount) Reconcile(request reconcile.Request) (reconcile.Resul
 			return reconcile.Result{Requeue: true}, r.Client.Update(context.TODO(), currentAcctInstance)
 		}
 
+		awsAssumedRoleClient, creds, err := r.handleCreateAdminAccessRole(reqLogger, currentAcctInstance, awsSetupClient)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
 		// Use the same ID applied to the account name for IAM usernames
 		iamUserUHC := fmt.Sprintf("%s-%s", iamUserNameUHC, currentAcctInstance.Labels[awsv1alpha1.IAMUserIDLabel])
 		secretName, err := r.BuildIAMUser(reqLogger, awsAssumedRoleClient, currentAcctInstance, iamUserUHC, request.Namespace)
@@ -430,14 +430,27 @@ func (r *ReconcileAccount) Reconcile(request reconcile.Request) (reconcile.Resul
 	}
 
 	if currentAcctInstance.IsReady() {
+		roleToAssume := getAssumeRole(currentAcctInstance)
+		awsClient, _, err := r.assumeRole(reqLogger, currentAcctInstance, awsSetupClient, roleToAssume, "")
+		if err != nil {
+			reqLogger.Error(err, "failed building BYOC client from assume_role")
+			// We don't want to error here as erroring will requeue and we will end in an
+			// infinite loop.  So we just log the error and exit.
+			// TODO maybe there's a better way to handle this?
+			return reconcile.Result{}, nil
+		}
 
 		// Check that secret is valid and reheal it if not
 		iamUserUHC := fmt.Sprintf("%s-%s", iamUserNameUHC, currentAcctInstance.Labels[awsv1alpha1.IAMUserIDLabel])
 
 		reqLogger.Info("probing account secret")
-		err := r.ProbeSecret(reqLogger, currentAcctInstance, awsAssumedRoleClient, iamUserUHC, request.Namespace)
+		err = r.ProbeSecret(reqLogger, currentAcctInstance, awsClient, iamUserUHC, request.Namespace)
 		if err != nil {
-			return reconcile.Result{}, err
+			reqLogger.Error(err, "failed to probe secret")
+			// We don't want to error here as erroring will requeue and we will end in an
+			// infinite loop.  So we just log the error and exit.
+			// TODO maybe there's a better way to handle this?
+			return reconcile.Result{}, nil
 		}
 	}
 
```

**Pull request merged**